src/interpreter.rs

```diff
@@ -120,3 +129,1207 @@ bitflags::bitflags! {
+            //
+            // Testing if this vin is not final is sufficient to
+            // prevent this condition. Alternatively we could test all
+            // inputs, but testing just this input minimizes the data
+            // required to prove correct CHECKLOCKTIMEVERIFY execution.
+        } else {
+            !self.is_final
+        }
+    }
+}
+
+pub fn verify_script(
+    script_sig: &Script,
+    script_pub_key: &Script,
+    flags: VerificationFlags,
+    checker: &dyn SignatureChecker,
+) -> Result<(), ScriptError> {
+    if flags.contains(VerificationFlags::SigPushOnly) && !script_sig.is_push_only() {
+        return set_error(ScriptError::SigPushOnly);
+    }
+
+    let mut stack = Stack(Vec::new());
+    let mut stack_copy = Stack(Vec::new());
+    if !eval_script(&mut stack, script_sig, flags, checker)? {
+        // serror is set
+        return set_error(ScriptError::UnknownError);
+    }
+    if flags.contains(VerificationFlags::P2SH) {
+        stack_copy = stack.clone()
+    }
+    if !eval_script(&mut stack, script_pub_key, flags, checker)? {
+        // serror is set
+        return set_error(ScriptError::UnknownError);
+    }
+    if stack.empty() {
+        return set_error(ScriptError::EvalFalse);
+    }
+    if !cast_to_bool(stack.back()?) {
+        return set_error(ScriptError::EvalFalse);
+    }
+
+    // Additional validation for spend-to-script-hash transactions:
+    if flags.contains(VerificationFlags::P2SH) && script_pub_key.is_pay_to_script_hash() {
+        // script_sig must be literals-only or validation fails
+        if !script_sig.is_push_only() {
+            return set_error(ScriptError::SigPushOnly);
+        };
+
+        // Restore stack.
+        swap(&mut stack, &mut stack_copy);
+
+        // stack cannot be empty here, because if it was the
+        // P2SH  HASH <> EQUAL  scriptPubKey would be evaluated with
+        // an empty stack and the `eval_script` above would return false.
+        assert!(!stack.empty());
+
+        let pub_key_serialized = stack.back()?.clone();
+        let pub_key_2 = Script(pub_key_serialized.as_slice());
+        stack.pop()?;
+
+        if !eval_script(&mut stack, &pub_key_2, flags, checker)? {
+            // serror is set
+            return set_error(ScriptError::UnknownError);
+        }
+        if stack.empty() {
+            return set_error(ScriptError::EvalFalse);
+        }
+        if !cast_to_bool(stack.back()?) {
+            return set_error(ScriptError::EvalFalse);
+        }
+    }
+
+    // The CLEANSTACK check is only performed after potential P2SH evaluation,
+    // as the non-P2SH evaluation of a P2SH script will obviously not result in
+    // a clean stack (the P2SH inputs remain).
+    if flags.contains(VerificationFlags::CleanStack) {
+        // Disallow CLEANSTACK without P2SH, as otherwise a switch CLEANSTACK->P2SH+CLEANSTACK
+        // would be possible, which is not a softfork (and P2SH should be one).
```

Comment:
That was Bitcoin Core's motivation, but it doesn't make sense for Zcash because P2SH was supported from genesis.
```suggestion
        // Disallow CLEANSTACK without P2SH, because Bitcoin did.
```